connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/LiteralProtoConverter.scala

```diff
@@ -0,0 +1,297 @@
+  def toLiteralProtoBuilder(literal: Any): proto.Expression.Literal.Builder = {
+    val builder = proto.Expression.Literal.newBuilder()
+
+    def decimalBuilder(precision: Int, scale: Int, value: String) = {
+      builder.getDecimalBuilder.setPrecision(precision).setScale(scale).setValue(value)
+    }
+
+    def calendarIntervalBuilder(months: Int, days: Int, microseconds: Long) = {
+      builder.getCalendarIntervalBuilder
+        .setMonths(months)
+        .setDays(days)
+        .setMicroseconds(microseconds)
+    }
+
+    def arrayBuilder(array: Array[_]) = {
+      val ab = builder.getArrayBuilder
+        .setElementType(componentTypeToProto(array.getClass.getComponentType))
+      array.foreach(x => ab.addElement(toLiteralProto(x)))
+      ab
+    }
+
+    literal match {
+      case v: Boolean => builder.setBoolean(v)
+      case v: Byte => builder.setByte(v)
+      case v: Short => builder.setShort(v)
+      case v: Int => builder.setInteger(v)
+      case v: Long => builder.setLong(v)
+      case v: Float => builder.setFloat(v)
+      case v: Double => builder.setDouble(v)
+      case v: BigDecimal =>
+        builder.setDecimal(decimalBuilder(v.precision, v.scale, v.toString))
+      case v: JBigDecimal =>
+        builder.setDecimal(decimalBuilder(v.precision, v.scale, v.toString))
+      case v: String => builder.setString(v)
+      case v: Char => builder.setString(v.toString)
+      case v: Array[Char] => builder.setString(String.valueOf(v))
+      case v: Array[Byte] => builder.setBinary(ByteString.copyFrom(v))
+      case v: collection.mutable.WrappedArray[_] => toLiteralProtoBuilder(v.array)
+      case v: LocalDate => builder.setDate(v.toEpochDay.toInt)
+      case v: Decimal =>
+        builder.setDecimal(decimalBuilder(Math.max(v.precision, v.scale), v.scale, v.toString))
+      case v: Instant => builder.setTimestamp(DateTimeUtils.instantToMicros(v))
+      case v: Timestamp => builder.setTimestamp(DateTimeUtils.fromJavaTimestamp(v))
+      case v: LocalDateTime => builder.setTimestampNtz(DateTimeUtils.localDateTimeToMicros(v))
+      case v: Date => builder.setDate(DateTimeUtils.fromJavaDate(v))
+      case v: Duration => builder.setDayTimeInterval(IntervalUtils.durationToMicros(v))
+      case v: Period => builder.setYearMonthInterval(IntervalUtils.periodToMonths(v))
+      case v: Array[_] => builder.setArray(arrayBuilder(v))
+      case v: CalendarInterval =>
+        builder.setCalendarInterval(calendarIntervalBuilder(v.months, v.days, v.microseconds))
+      case null => builder.setNull(nullType)
+      case _ => unsupported(s"literal $literal not supported (yet).")
+    }
+  }
+
+  /**
+   * Transforms literal value to the `proto.Expression.Literal`.
+   *
+   * @return
+   *   proto.Expression.Literal
+   */
+  def toLiteralProto(literal: Any): proto.Expression.Literal =
+    toLiteralProtoBuilder(literal).build()
+
+  private def componentTypeToProto(clz: Class[_]): proto.DataType = clz match {
```

Comment:
Can we simplify this a bit. Please build the conversion from `Class[_] -> org.apach.spark.sql.types.DataType`. Then call the DataTypeProtoConverter to convert it to its proto equivalent.

Reply:
OK

Reply:
So I'd slightly pull it apart more. Create one method that convers the `Class` into a `DataType`, and another converts it into `proto.DataType`. For example:
```scala
private def toDataType(cls: Class[_]): DataType = {
  ... // most of your current code
}

provate def toProtoDataType(cls: Class[_]): proto.DataType = {
  toConnectProtoType(toDataType(cls))
}
```
That way it is easier to test the individual components. Bonus point if you write a UT here.

Reply:
hmm .. should we make 

https://github.com/apache/spark/blob/4b8b36834e2902184088d171ee919cf5c5e3b9a3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L109-L148

as public ? I think it It is equivalent to `toDataType `

Reply:
Let's keep this function. If we implement MapType and other types, should be code changes here